R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+    new_name <- value[names(x) == old_name]
+
+    # Change 2nd-tier first
+    for (sub_name in names(old_entry)) {
+      names(x[[sub_name]])[names(x[[sub_name]]) == old_name] <- new_name
+    }
+
+    # Change in first tier
+    names(x)[names(x) == old_name] <- new_name
+  }
+  class(x) <- c("join_keys", "list")
+  x
+}
+
+#' @rdname join_keys
+#' @details
+#' Getter for `join_keys` that returns the relationship between pairs of datasets.
+#'
+#' @param join_keys_obj (`join_keys`) object to extract the join keys
+#' @param dataset_1 (`character`) name of first dataset.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Getter for join_keys ----
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]][["ds2"]] <- "some_col"
+#' jk[["ds1"]][["ds3"]] <- "some_col2"
+#'
+#' jk["ds1"]
+#' jk[1:2]
+#' jk[c("ds1", "ds2")]
+`[.join_keys` <- function(join_keys_obj, dataset_1 = NULL) {
+  # Protection against missing being passed through functions
+  if (missing(dataset_1)) dataset_1 <- NULL
+
+  if (is.null(dataset_1)) {
+    return(join_keys_obj)
+  }
+
+  if (checkmate::test_integerish(dataset_1) || checkmate::test_logical(dataset_1)) {
+    dataset_1 <- names(join_keys_obj)[dataset_1]
+  }
+
+  # When retrieving a relationship pair, it will also return the symmetric key
+  new_jk <- new_join_keys()
+  for (ix in dataset_1) {
+    new_jk[[ix]] <- join_keys_obj[[ix]]
+  }
+  common_parents_ix <- names(parents(join_keys_obj)) %in% names(new_jk)
+  if (any(common_parents_ix)) parents(new_jk) <- parents(join_keys_obj)[common_parents_ix]
+
+  new_jk
+}
+
+#' @rdname join_keys
+#' @details
+#' Setter via index directly (bypassing the need to use `join_key()`).
+#' When `dataset_2` is omitted, it will create a primary key with `dataset_2 = dataset_1`.
+#'
+#' @param value (`character` vector) value to assign.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Setter via index ----
+#'
+#' jk <- join_keys(
+#'   join_key("ds1", "ds2", "col12"),
+#'   join_key("ds3", "ds4", "col34")
+#' )
+#'
+#' # overwrites previously defined key
+#' jk["ds1"] <- list(ds2 = "(new)co12")
+#' jk["ds1"] <- list(ds3 = "col13", ds4 = "col14")
+#' jk
+#'
+#' jk[c("ds1", "ds2")] <- list(ds5 = "col*5")
+#' jk[c(1, 2)] <- list(ds5 = "col**5")
+#'
+#' # Creates primary key by only defining `dataset_1`
+#' jk["ds1"] <- "primary_key"
+#' jk
+`[<-.join_keys` <- function(join_keys_obj, dataset_1, value) {
+  checkmate::assert(
+    combine = "or",
+    checkmate::check_character(dataset_1),
+    checkmate::check_integerish(dataset_1)
+  )
+
+  if (checkmate::test_integerish(dataset_1)) {
+    dataset_1 <- names(join_keys_obj)[dataset_1]
+  }
+
+  checkmate::assert(
+    combine = "or",
+    checkmate::check_character(value),
+    checkmate::check_list(value, names = "named", types = "character", null.ok = TRUE)
+  )
+
+  # Assume characters as being primary keys
+  if (checkmate::test_character(value)) {
+    value <- lapply(dataset_1, function(dataset_ix) {
+      value
+    })
+    names(value) <- dataset_1
```

Comment:
Seems a little strange to use `lapply` where `FUN` does nothing.
Did you mean something like this?
```suggestion
    value <- rep(value, length(dataset_1)
    value <- structure(value, names = value)
```

You can also get the `names` with `sapply(..., simplify = FALSE)`, by the way.

Reply:
Silly piece of code, this was already corrected. Thanks :bow: